tools/validators/instance_validator/validate/entity_instance.py

```diff
@@ -123,13 +127,72 @@ def _ValidateLinks(self):
+      target_entity_type = self.universe.GetEntityType(self.namespace,
+                                                       self.type_name)
+      all_fields_dict = target_entity_type.GetAllFields().copy()
+      for source_field, target_field in fields_map.items():
+        # check the fields are present
+        # assumes that the namespace is `` for now
+        if not src_entity_type.HasField('/'+source_field.data):
+          print('Invalid link field source: ', source_field)
+          return False
+        if not target_entity_type.HasField('/'+target_field.data):
+          print('Invalid link field target: ', target_field)
+          return False
+        # the found field is removed from the dictionary,
+        # once all the links are visited,
+        # check if there are fields that are not linked
+        # the dict fields keys are: namespace/fields, the namespace is empty.
+        all_fields_dict.pop('/' + target_field.data, None)
+        # check the units are matching
+        valid_units_src = src_entity_instance.universe\
+          .GetUnitsMapByMeasurement(source_field.data)
+        valid_units_target = self.universe\
+          .GetUnitsMapByMeasurement(target_field.data)
+        if valid_units_src != valid_units_target:
+          print('Links dont have the same units: ',
+                valid_units_src, valid_units_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+        if valid_units_src is not None and\
+            set(valid_units_src) != set(valid_units_target):
+          print('Links dont have the same units: ',
+                valid_units_src, valid_units_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+	      # todo: discuss checks on the measurement descriptors
+        # check the states are matching
+        valid_states_src = src_entity_instance.universe\
+          .GetStatesByField(source_field.data)
+        valid_states_target = self.universe\
+          .GetStatesByField(target_field.data)
+        if valid_states_src != valid_states_target:
+          print('Links dont have the same states: '
+                , valid_states_src, valid_states_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+        if valid_states_src is not None and\
+            set(valid_states_src) != set(valid_states_target):
+          print('Links dont have the same states: ',
+                valid_states_src, valid_states_target)
+          print('Links error for the following fields (source, target): '
+                , source_field, target_field)
+          return False
+      # check if the rest of the links not included are optional
+      for optional_field_name in all_fields_dict.values():
```

Comment:
nit: call it field_name you don't know it's optional yet.

Reply:
goo dpoint